### ISSUE TITLE
Game conversion fails from 3.0.6 to 3.1: Can't change state whilst flushing queries

### ISSUE THREAD

**Issue by nathanjwtx** [godotengine/godot#25499]
Possibly related to https://github.com/godotengine/godot/issues/24694

**Godot version:**
3.1 Beta 2 & 3

**OS/device including version:**
Windows 10 i5 NVidia 1050GTX

**Issue description:**
Converting a project that works in 3.0.6 starts crashing

```
E 0:00:12:0238   Can't change this state while flushing queries. Use call_deferred() or set_deferred() to change monitoring state instead
  <C Source>     servers/physics_2d/physics_2d_server_sw.cpp:630 @ body_set_mode()
  <Stack Trace>  :0 @ Godot.Node Godot.NativeCalls.godot_icall_1_504(IntPtr , IntPtr , Int32 )()
                 PackedScene.cs:94 @ Godot.Node Godot.PackedScene.Instance(Godot.PackedScene+GenEditState )()
                 Main.cs:154 @ void Main.SpawnRock(Int32 , System.Nullable`1[Godot.Vector2] , System.Nullable`1[Godot.Vector2] )()
                 Main.cs:171 @ void Main._on_Rock_Boom(Int32 , Single , Godot.Vector2 , Godot.Vector2 , Boolean )()
                 :0 @ System.Object Godot.NativeCalls.godot_icall_2_516(IntPtr , IntPtr , System.String , System.Object[] )()
                 Object.cs:326 @ System.Object Godot.Object.EmitSignal(System.String , System.Object[] )()
                 Rock.cs:87 @ void Rock.Explode(Boolean )()
                 Bullet.cs:35 @ void Bullet._on_Area2D_body_entered(Godot.Object )()
```

**Steps to reproduce:**
Open project below, start the game and shoot an asteroid.

**Minimal reproduction project:**
https://github.com/nathanjwtx/space-rocks-3-1

**Comment by eon-s** [comment]
Do not forget to mention when is a Mono problem so it gets properly tagged.
--------------------------------
Have you tried fixing the error as the log says?
`Use call_deferred() or set_deferred() to change monitoring state instead`

It may happen when you change a CollisionObject property during a signal (and it seems that you call some method to change properties during signals).

**Comment by nathanjwtx** [comment]
@eon-s sorry about the Mono tag.

I haven't tried doing what the log suggests as I don't fully understand it never having used either of those methods before. This seems like a substantial breaking change for my project which might be better off left in 3.0.6 if I can't easily figure this out.

**Comment by nathanjwtx** [comment]
I don't know, it never occurred when running under 3.0.6 and C#. I'm not familiar with GDScript so I couldn't say for sure although this project is based on a GDScript project.

The signal call back being the "_on_MySignal" type method?

**Comment by eon-s** [comment]
yes, whatever function is on the connection, I do not know the call deferred syntax for C# but should be similar.

**Comment by nathanjwtx** [comment]
Changing to using CallDeferred then raised another problem. Godot 3.1 doesn't like nullable Vector2s. I don't know if this was intentional or not in the move from 3.0.6

   

```
 private void SpawnRock(int size, Vector2? pos = null, Vector2? velocity = null)
    {
        if (pos == null)
        {
            GetNode<PathFollow2D>("RockPath/RockSpawn").SetOffset(_random.Next(0, int.MaxValue));
            pos = GetNode<PathFollow2D>("RockPath/RockSpawn").Position;
        }

        if (velocity == null)
        {
            Vector2 temp = new Vector2(1, 0);
            velocity = temp.Rotated(_random.Next(0, 360)) * _random.Next(100, 150);
        }
        var r = (Rock) RockScene.Instance();
        r._screensize = _screenSize;
        r.Start((Vector2) pos, (Vector2) velocity, size, _random.Next(0, 4));
        GetNode<Node>("Rocks").AddChild(r);
        r.Connect("Boom", this, "_on_Rock_Boom");
    }
```

**Comment by eon-s** [comment]
Just a guess here but try `= default` instead? 
Nullables have some peculiarities, check referenced issue above.

**Comment by nathanjwtx** [comment]
with @paulloz help I fixed the issue by creating an additional method which in turn calls an overloaded method. Do you want the issue closed or left open for reference?

**Comment by eon-s** [comment]
I think that @neikeq may know if this is a bug or should be done that way and be documented instead.

**Comment by nathanjwtx** [comment]
Does this mean that it isn't Mono specific?

**Comment by Xrayez** [comment]
@nathanjwtx it happened to me using GDScript alone, so most likely yes, it's not language specific.

**Comment by nathanjwtx** [comment]
Ah, that was due to the Mono version not correctly recognising Vector2s being nullable. I had to create the method with one parameter and no Vector2 args, then the overloaded one with Vector2s.

**Comment by nathanjwtx** [comment]
But I also had to use CallDeferred in a couple of other places when calling methods on other classes

**Comment by exts** [comment]
@eon-s yeah this was something reduz added here https://github.com/godotengine/godot/commit/0b1e93ccd43815d93de031b4f23fbfcd9ae7a34c#diff-ea3eacd38c73d9bf9b339a2c2a7dd2e3

**Comment by neikeq** [comment]
What's the issue left here? As we were discussing, this is expected behavior in Godot and the solution is to defer the code execution for later (with `CallDeferred` for example).
Is there any Mono specific issue? There is #19868 for the nullable issue.

**Comment by nathanjwtx** [comment]
@neikeq no i don't think there is a mono specific issue any longer. Most likely needs to be documented that CallDeferred() is now required whereas before it was optional.

**Comment by akien-mga** [comment]
Related to #24694 and #26148.

**Comment by akien-mga** [comment]
The reproduction project is no longer available though.

**Comment by akien-mga** [comment]
Fixed by b84b015.
